### PR TITLE
Add rust-based CI tool

### PR DESCRIPTION
**Pull request by vrutkovs**
Create a small Rust app which reuses Cincinnati code to verify PRs.

The following checks are implemented:
* Verify files in `channels` and `blocked-edges` are valid yamls, no 
other files are present.
* Check that all files can be deserialized in Cincinnati objects
* Verify that all mentioned releases are pushed to Quay

This will not be hooked up to CI yet, as it requires openshift/release PR to replace python hack.

TODO:
* [x] Requires https://github.com/openshift/cincinnati/pull/304 to merge changes from Cincinnati fork.

/cc @wking @steveeJ

### CODE REVIEW

File: hack/src/check_releases.rs
Comment (steveeJ) [inline hack/src/check_releases.rs:37]: I think we can be more efficient here by comparing two sets. (Typing in the github comment window here without actually verifying the code, please beware :wink:)

```suggestion
  .context("failed to fetch all release metadata")?
  .into_iter()
  .map(|r| r.metadata.version)
  .collect::<HashSet<Version>>();

  println!("Verifying all releases are uploaded");
  // assume we've assigned the above HashSet to `released_versions`
  let missing_versions = found_versions - released_versions;
  if !missing_versions.is_empty() {
    bail!("Missing the following versions in scraped images: {:?}", missing_versions)
  }
```

File: hack/src/verify_yaml.rs
Comment (steveeJ) [inline hack/src/verify_yaml.rs:28]:
```suggestion
  let channel_path = data_dir.join(plugin::CHANNELS_DIR).canonicalize()?;
```

File: hack/src/verify_yaml.rs
Comment (steveeJ) [inline hack/src/verify_yaml.rs:15]:
```suggestion
  let extension_re = Regex::new("ya+ml")?;
```

**Review by steveeJ — COMMENTED** [review]
Thanks for this work! What do you think of creating a sub-directory for the crate separate from _hack_?

File: hack/src/verify_yaml.rs
Comment (vrutkovs) [reply hack/src/verify_yaml.rs]: Its probably easier to construct error message here, wdyt?

**Comment by vrutkovs** [comment]
>What do you think of creating a sub-directory for the crate separate from hack

I'm not sure what's the Rust standard for this, but yeah, lets do so

File: hack/src/main.rs
Comment (vrutkovs) [reply hack/src/main.rs]: Removed extra `pub`s

File: hack/src/check_releases.rs
Comment (vrutkovs) [reply hack/src/check_releases.rs]: Good idea, implemented this using `HashSet::Difference`

File: hack/src/verify_yaml.rs
Comment (steveeJ) [inline hack/src/verify_yaml.rs:41]: I think that we don't actually need the `ProcessingError` enum if we just coerce everything to a string at the end. It would in fact be enough if you just appended the error messages to a mutable string throughout the processing. I don't feel strongly about it though so I'm not opposed to keeping this either.

**Review by steveeJ — COMMENTED** [review]
Are you still deciding whether you move to the crate into its own directory or have you decided not to?

File: hack/Cargo.toml
Comment (wking) [inline hack/Cargo.toml:5]: Is this some Cargo magic, or should it be 2020?

File: hack/src/verify_yaml.rs
Comment (wking) [inline hack/src/verify_yaml.rs:22]: Do we need local walkers for validation?  I was expecting us to be able to wash through [the `github-secondary-metadata-scrape` plugin][1].

[1]: https://github.com/openshift/cincinnati/tree/3f5ea5804d1fe67f657e900d652c75622480f9ab/cincinnati/src/plugins/internal/graph_builder/github_openshift_secondary_metadata_scraper

Comment (wking) [reply hack/src/verify_yaml.rs:22]: Somewhat in this space, and feel free to punt as out of scope for a v1 PR, but it would be nice to be able to feed this tool graph-builder and policy-engine plugin configs, which we could copy/edit/paste from our production update service deployment, and have it ingest the local graph-data checkout, wash through the rest of the plugin stack, and spit out assorted policy-engine output graphs for all architectures and channels.  Or diff vs. the current production graph.  Or something...

File: hack/src/verify_yaml.rs
Comment (vrutkovs) [reply hack/src/verify_yaml.rs:41]: I think it would be useful. We might want to add more error types later - i.e. signature check - with customized Display

File: hack/src/verify_yaml.rs
Comment (vrutkovs) [reply hack/src/verify_yaml.rs:22]: metadata-scrape is too forgiving, it intentionally skips files with no extension etc. Perhaps later we could introduce "strict" mode there and merge the walkers.

>it would be nice to be able to feed this tool graph-builder and policy-engine plugin configs

That would be useful to setup configuration (if the registry or org name changes). Probably out of scope now

> spit out assorted policy-engine output graphs for all architectures and channels

that could be done, but its not clear what's the expected graph is. Every change would produce a diff, but its not trivial to validate it. Its out of scope for v1, but might become useful

File: hack/Cargo.toml
Comment (vrutkovs) [reply hack/Cargo.toml:5]: Rust has two major editions - [2015 and 2018](https://doc.rust-lang.org/edition-guide/editions/index.html)

File: hack/src/verify_yaml.rs
Comment (steveeJ) [reply hack/src/verify_yaml.rs:41]: Adding more types doesn't make it more useful if we still only use it for string coercion :wink: What would convince me if we pass back the error vector to the caller, which would allow it to use the type system to figure out error conditions.

File: hack/src/verify_yaml.rs
Comment (steveeJ) [reply hack/src/verify_yaml.rs:23]: @wking I would've gone in a similar direction to yours but I never got around to pursuing the implementation. My idea was to make a CLI tool very similar to what the GB and PE are today, but without the looping and serving aspects. It would just run the plugin chain once and yield the graph output. The latter could be handled via a plugin as. It might actually be pretty easy to add a config option such as `runonce` to the graphbuilder and add a small plugin which emits the graph to disk, stdout, or whatever we desire.

File: hack/src/verify_yaml.rs
Comment (vrutkovs) [reply hack/src/verify_yaml.rs:41]: >Adding more types doesn't make it more useful 

The idea is to leave formatting errors to Debug impl of ProcessingError. Other tests may extend that struct later.

> What would convince me if we pass back the error vector to the caller

`verify_yamls::run` may throw generic errors, e.g. directory not readable or "blocked-edges" dir cannot be canonicalized.

File: hack/src/verify_yaml.rs
Comment (steveeJ) [reply hack/src/verify_yaml.rs:23]: @wking I came up with https://github.com/openshift/cincinnati/pull/307 which allows the removal of the walker and error types in this PR. @vrutkovs and I decided that this is enough for this PR and would like to postpone further optimization. Is that fine with you?

Comment (wking) [reply hack/src/verify_yaml.rs:23]: I am fine punting anything.

File: hack/Dockerfile
Comment (steveeJ) [inline hack/Dockerfile:5]: I don't understand when this switches to the `hack/` directory

**Review by steveeJ — CHANGES_REQUESTED** [review]
Given that we'll further rework this to reuse cincinnati plugin system I think that we're almost good to merge this initial version!

Only two things that I think need changing before we can merge:
 
* It looks like the rust code hasn't run through rustfmt.
* The rust crate is lumped together with the other files in the _hack_ directory.

Could we run the python tool and the rust tool side-by-side for a while to see if they behave the same?

File: hack/Dockerfile
Comment (vrutkovs) [reply hack/Dockerfile:5]: This runs within `hack/` directory:
```
- context_dir: hack
  dockerfile_path: Dockerfile
  from: rust-toolset
  to: publish
```
in `release` repo

**Comment by vrutkovs** [comment]
Pushed a fix which formats the code using rustfmt (I'll add a new test to verify that in `release` repo)

**Comment by steveeJ** [comment]
> We'll have to live with that for some time until CI is switched to Rust tool only.

I don't understand why that is. It seems like we're in control over all systems here so we can adjust anything as needed. Would you mind to elaborate?

**Comment by vrutkovs** [comment]
After we merge this tool other PRs might be created, so we can't remove python scripts right away. Also, when `release` PR is created some errors might pop up, so some fixed to this repo might be required. 

Hopefully it would take a day or two to remove all python scripts, so I wouldn't worry about current state too much

**Comment by steveeJ** [comment]
> After we merge this tool other PRs might be created, so we can't remove python scripts right away. Also, when `release` PR is created some errors might pop up, so some fixed to this repo might be required.

I didn't mean to remove the _hack_ directory altogether, just to put the Rust crate in a separate, possibly nested, directory. Wouldn't that work?

**Comment by vrutkovs** [comment]
Why would we want to move it out of `hack`? Eventually it would contain rust pkg only.

I don't mind creating a new dir for that - do you have any catchy name for it in mind?

**Comment by steveeJ** [comment]
> I don't mind creating a new dir for that - do you have any catchy name for it in mind?

Is `graph-data.rs` catchy enough? :smile:

File: graph-data.rs/Dockerfile
Comment (steveeJ) [inline graph-data.rs/Dockerfile:5]: From the help output there seems to be missing a `PATH`

```
        --path <PATH>                                Filesystem path to local crate to install
```

File: graph-data.rs/src/main.rs
Comment (steveeJ) [inline graph-data.rs/src/main.rs:12]:
```suggestion
    let mut runtime = tokio::runtime::Runtime::new()?;
```

**Review by steveeJ — COMMENTED** [review]
This LGTM now! It would be nice to see the CI job in action. What do you think of hooking it up before we merge this PR and marking it as optional so that it doesn't block other PRs?

**Comment by vrutkovs** [comment]
it can't be hooked up before we merge - rehearsals will fail. Lets merge this so that I could land a preliminary version of this and make it "always_run: true" but "optional:true"?

**Comment by steveeJ** [comment]
> it can't be hooked up before we merge - rehearsals will fail. Lets merge this so that I could land a preliminary version of this and make it "always_run: true" but "optional:true"?

/lgtm

**Comment by steveeJ** [comment]
Giving you the opportunity to squash:

/hold

**Review by steveeJ — APPROVED** [review]
/lgtm

**Pull request merged**
